### PR TITLE
Verify foreign keys after loading fixtures

### PR DESCRIPTION
**Pull request by ghiculescu**
When writing fixtures, it's possible to define associations that don't exist, even if a foreign key exists. For example:

```yml
# test/fixtures/parrots.yml
george:
  name: "Curious George"
  pirate: redbeard

# test/fixtures/pirates.yml
blackbeard:
  name: "Blackbeard"
 ```

When the fixtures are created, `parrots(:george).pirate` will be `nil`, but it's not clear why. This can make it hard to debug tests and can give false confidence in passing ones.

This happens because Rails [disables referential integrity](https://github.com/rails/rails/blob/f263530bf709f611920b5f663882e5113b4f984b/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L407) when inserting fixtures. This makes the fixtures algorithm much simpler - it can just create the fixtures in alphabetical order and assume that the other side of a foreign key constraint will *eventually* be added.

Ideally we would check foreign keys once all fixtures have been loaded, so that we can be sure that the foreign key constraints were met. This PR introduces that. To enable it:

```ruby
config.active_record.verify_foreign_keys_for_fixtures = true
```

I'm proposing we enable this in 7.0 for new apps and have added it to new framework defaults. When run against our app, it found 3 fixture files with unmet FK constraints - turns out all those fixtures weren't being used and were safe to delete.

The implementation works with SQLite (which has a builtin pragma for this) and PostgreSQL (which we use, and it was easy enough to write a query for). I haven't implemented for MySQL as I wasn't able to find any good advice on how to do it, but would love to add it in if anyone knows how.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb [hunk 1]
```diff
@@ -211,6 +211,11 @@ def disable_referential_integrity # :nodoc:
         end
       end
 
+      def verify_all_foreign_keys! # :nodoc:
+        fks = execute("PRAGMA foreign_key_check")
+        raise "Foreign key violations found!" unless fks.blank?
```
Comment (ghiculescu) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb:216]: This error message could be a lot better...

File: activerecord/lib/active_record/connection_adapters/postgresql/referential_integrity.rb [hunk 2]
```diff
@@ -37,6 +37,36 @@ def disable_referential_integrity # :nodoc:
...
+            END;
+            $$;
+          SQL
+
+          begin
+            transaction(requires_new: true) do
+              execute(sql)
+            end
+          rescue ActiveRecord::StatementInvalid
+            raise "Foreign key violations found!"
```
Comment (ghiculescu) [inline activerecord/lib/active_record/connection_adapters/postgresql/referential_integrity.rb:67]: This error message could be a lot better...

File: guides/source/configuring.md [hunk 3]
```diff
@@ -481,6 +481,13 @@ in controllers and views. This defaults to `false`.
 
 * `config.active_record.enumerate_columns_in_select_statements` when true, will always include column names in `SELECT` statements, and avoid wildcard `SELECT * FROM ...` queries. This avoids prepared statement cache errors when adding columns to a PostgreSQL database for example. Defaults to `false`.
 
+* `config.active_record.destroy_all_in_batches` ensures
```
Comment (ghiculescu) [inline guides/source/configuring.md:484]: Not related to this PR. Moved this to the right section, it was under the Schema Dumper config which was incorrect.

File: activerecord/test/cases/fixtures_test.rb [hunk 4]
```diff
@@ -790,6 +790,76 @@ def test_number2
...
+    # this creates FK violations since Parrot and ParrotTreasure records are created.
+    # those violations can cause false positives in these tests. since they aren't related to these tests we
+    # delete the irrelevant records here (this test is transactional so it's fine).
+    Parrot.all.each(&:destroy)
+  end
+
+  def test_raises_fk_violations
+    fk_pointing_to_non_existent_object = <<~FIXTURE
+  first:
+    fk_object_to_point_to: one
```
Comment (rafaelfranca) [inline activerecord/test/cases/fixtures_test.rb:814]:
```suggestion
      first:
        fk_object_to_point_to: one
```

File: activerecord/lib/active_record/connection_adapters/postgresql/referential_integrity.rb [hunk 5]
```diff
@@ -37,6 +37,37 @@ def disable_referential_integrity # :nodoc:
...
+                EXECUTE (r.constraint_check);
+              END LOOP;
+            END;
+            $$;
+          SQL
+
+          begin
+            transaction(requires_new: true) do
+              execute(sql)
+            end
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/postgresql/referential_integrity.rb:65]:
```suggestion
            end

```

File: activerecord/lib/active_record/connection_adapters/postgresql/referential_integrity.rb [hunk 6]
```diff
@@ -37,6 +37,37 @@ def disable_referential_integrity # :nodoc:
           rescue ActiveRecord::ActiveRecordError
           end
         end
+
+        def all_foreign_keys_valid? # :nodoc:
+          sql = <<~SQL
+            do $$
+              declare r record;
+            BEGIN
+            FOR r IN  (
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/postgresql/referential_integrity.rb:46]:
```suggestion
            FOR r IN (
```

**Comment by ghiculescu** [comment]
PSA: if anyone who knows more MySQL than me comes across this it’d be great to see it implemented for that too. I’m happy to help with the PR

**Comment by zzak** [comment]
@ghiculescu If it's worth adding this to `thisweekinrails` do you think it's worth adding a changelog? :bow:

**Comment by ghiculescu** [comment]
Yes. 🤦  I will tomorrow.

**Comment by dvodvo** [comment]
` I'm not confident that's something we want to have in Rails.`
I can see why this would tend to be very long and agree with this conclusion.  
I would even add that launching a Unit test should make no assumptions about fixtures; that appears to be the wrong place for it.

However, sometimes the process of fishing out the guilty party when one **absolutely** needs many fixtures becomes excessively long.  Geometrically long, in fact.

Instead, when this monster rears its head, I would instinctively like to do something akin to: `rails test [path_to_fixtures][optionally: '/specific_fixtures.yml'`
The process would run through the invoked YAML files and stop upon the first error, reporting that.

**Comment by ghiculescu** [comment]
@mjc-gh @dvodvo you might be interested in this PR that @danini-the-panini is working on: https://github.com/rails/rails/pull/44943

**Comment by hey-leon** [comment]
@ghiculescu cool feature!

**Pull request merged**